proposals/roles.md

```diff
@@ -0,0 +1,665 @@
+TODO
+
+### Operators
+
+TODO
+User-defined conversion should be allowed, except where it conflicts with a built-in
+conversion (such as with an underlying type).  
+
+### Method group conversions
+
+https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/conversions.md#108-method-group-conversions
+TODO A single method is selected corresponding to a method invocation, 
+but with some tweaks related to normal form and optional parameters.  
+TODO There's also the scenario where a method group contains a single method (lambda improvements).  
+
+## Implementation details
+
+Roles are implemented as ref structs.  
+If the role any instance member, then we emit a ref field (of underlying type)
+into the ref struct and a constructor.  
+Values of role types are left as values of the underlying value type, until a role
+member is accessed. When a role member is accessed, a role instance is created
+with a reference to the underlying value and the member is accessed on that instance.
+
+```
+Role r = default(UnderlyingType); // emitted as a local of type `UnderlyingType`
+r.RoleMember(); // emitted as `new Role(ref r).RoleMember();`
+```
+
+Roles appearing in signatures are emitted as the role's underlying type
+marked with a modopt of the role type.
+
+```
+void M(Role r) // emitted as `void M(modopt(Role) UnderlyingType r)`
+```
+
+TODO how do we emit an extension type versus a handcrafted ref struct?  
+TODO how do we emit the relationship to underlying type? base type or special constructor?  
+TODO our emit strategy should allow using pointer types and ref structs as underlying types
+in the future.  
+TODO issues in async code with ref structs
+
+## Phase A: Adding static constants, fields, methods and properties
+
+In this first subset of the feature, the syntax is restricted to *extension_declaration*
+and containing only *constant_declaration* members and static *field_declaration*, 
+*method_declaration*, *property_declaration* and *type_declaration* members.  
+TODO: events?
+
+## Phase B. Roles and extensions with members
+
+In this second subset of the feature, the *role_declaration* becomes allowed
+and non-static members other than fields or auto-properties become allowed.
+
+### Role and extension type members
+
+The restrictions on modifiers from phase A remain (`new`).  
+Non-static members become allowed in phase B.  
+
+#### Fields
+
+A *field_declaration* in a *role_declaration* or *extension_declaration* 
+shall explicitly include a `static` modifier.  
+
+#### Methods
+
+TODO allow `this` (of type current role).  
+
+#### Properties
+
+Auto-properties must still be static (since instance fields are disallowed).  
+TODO allow `this` (of type current role).  
+
+#### Operators
+
+##### Conversions
+
+TODO
+```
+role R : U { } 
+R r = default;
+object o = r; // what conversion is that? if R doesn't have `object` as base type. What about interfaces?
+```
+
+Should allow conversion operators. Extension conversion is useful. 
```

Comment:
![neo going through the looping train station](https://thumbs.gfycat.com/ConstantCreepyHackee-max-1mb.gif)

Reply:
I have been reviewing this for a few hours so the sense of deja vu is getting strong.. 😉